keps/2724-topology-aware-schedling/README.md

```diff
@@ -0,0 +1,569 @@
+landing in the same topology) and is about to start we modify the corresponding
+PodTemplates in the Job object to inject the `kueue.x-k8s.io/topology`
+scheduling gate. For homogenous PodSet assignments it is enough to inject the
+corresponding nodeSelector in the PodTemplates.
+
+Then, there is a new component, called `TopologyUngater`, which is a Workload
+reconciler which lists all pods for a given TAS PodSet, and ensures that the
+pods in the expected number are un-gated to a given value.
+
+Along with the scheduling gate, to each pod template the
+`kueue.x-k8s.io/workload` and `kueue.x-k8s.io/podset` labels / annotations are
+added to facilitate quick lookup (List request) for all pods corresponding to
+the workload (and more specifically PodSetAssignment) by `TopologyAssigner`.
+
+The TopologyUngater watches for pod events which trigger the reconciliation
+loop. The reconciliations are batched by 1s periods to make sure multiple
+created pods at the similar time don't trigger too many reconciliations.
+
+In order to prevent ungating more pods as expected for a PodSet we consider to
+use the expectations mechanism. The expectations are set for when we are about
+to ungate a Pod. The expectation is fulfield if the Pod is observed as ungated
+or the ungating request fails. We hold ungating if there are pending ungatings
+within the PodSet.
+
+### Support for ReplicatedJobs in JobSet
+
+Currently, in the PodSet we just keep the number of Pods created for a given
+PodTemplate. However, in case of JobSet the pods are naturally split between
+instances of replicated Jobs.
+
+Thus, without any extra work for JobSet, if JobSet has more than one
+ReplicatedJob its Pods could thus be assigned by TAS to different topologies,
+making the assignment suboptimal.
+
+In order to improve accuracy of the support for ReplicatedJobs we do the
+following API adjustments:
+
+```golang
+type PodSet struct {
+  ...
+
+  // ReplicatedJobCount indicates the number of replicated Jobs being span by
+  // the PodSet. Each replicated Job is assumed to have the same number of Pods
+  // with the same template.
+  // Default: 1
+  ReplicatedJobCount *int32
+
+  // ReplicatedJobKeyLabel specifies the name of the label which indicates
+  // the specific Job instance among ReplicatedJobs.
+  ReplicatedJobKeyLabel *string
+}
+```
+
+In case of JobSet `ReplicatedJobKeyLabel` could be either
+`jobset.sigs.k8s.io/job-index` or `jobset.sigs.k8s.io/replicatedjob-name` as
+both will uniquely identify a specific Job among the set of ReplicatedJob.
+
+```golang
+type PodSetAssignment struct {
+
+  // ReplicatedJobKey indicates the value of the Pod label which is used to
+  // specify a specific Job among a set of ReplicatedJobs.
+  ReplicatedJobKey *string
+}
+```
+Then, when the compute the assignments per pods sets corresponding to
+ReplicatedJobs, rather than the entire PodSet. Finally, the `PodSetUngater`
+ungates pods per replicatedJob assignment.
```

Comment:
Don't we also need changes to the `TopologyAssignment`?

How can the user specify that they want each replicated job to be in a particular topology domain?

Reply:
> Don't we also need changes to the `TopologyAssignment`?

I don't think this is needed as the TopologyAssignment contains a list of PodSetAssignments. If there are 2 ReplicatedJobs, there would be two items on the list. Alternative API is to have ReplicatedJobAssignment which would have the `ReplicatedJobKey` and the `PodSetAssignment`. However, we would need to do it this way day 0. So, I proposed the incremental approach.

> How can the user specify that they want each replicated job to be in a particular topology domain?

A user can use "require: rack" or "prefer: rack". In case of "require: rack" we will require that each instance of ReplicatedJob is fully contained within a rack, but there might be two jobs landing on the same rack. When using "prefer" a single ReplicatedJob may span 2 or more racks, but we try to compactify.

Reply:
So, in any other job, if I say `require: rack`, it means that the entire job will land on one rack.

But in jobset, it means that each replica will land on one rack. Or is the annotation only going to be in the Job templates? That's not clearly explained.

Reply:
This is related to the other question about the scope of topology. There are actually 3 possible levels where we want to require/prefer a topology domain:
L1. Workload level (comprises multiple PodSets for MPIJob or JobSet)
L2. PodSet  (comprises multiple ReplicatedJobs in case of JobSet)
L3. ReplicatedJob

I discussed with @mwielgus that all are valid use cases we may want to do at some point. The open questions are:
Q1: what are the APIs to request topology at every such level
Q2: which we prioritize for Alpha and which we can defer for Beta

Regarding Q1 I think we can have:
L1: "preferred-topology / required-topology" annotations at the workload level
L2: "replicaset-preferred-topology / replicaset-required-topology" annotations at the PodTemplate level
L3: "replica-preferred-topology / replica-required-topology" annotations at the PodTemplate level

I'm open to other/better names. Marcin suggests to start with whatever is simplest to deliver. I think looking at how flavorassigner works in Kueue L2 is the simplest, even if not the most functional. WDYT @alculquicondor @tenzen-y ? I will update the KEP accordingly.

Reply:
I think we should start with the most useful scenario, one that we know is already in use. And that's L3.

Reply:
I agree that L3 is probably the most useful. However at the same time it is significantly more complex than L2, on which Kueue already works right now.

L3 requires some non-trivial refactoring. It has to be done but for the first step I would suggest doing the most basic functionality to make sure that the thing works, more or less, E2E and **THEN** focusing on refactoring to enable the most valuable use cases.

Reply:
Ok, please update user stories and design accordingly.

Reply:
sgtm, the [Support for ReplicatedJobs in JobSet](https://github.com/kubernetes-sigs/kueue/pull/2725/files#diff-40027534eab1ffd244e3c960acbf91b55b9556bebd8cd56077ba5f3270648284R512) section focuses L3, but it still has some open questions. 

The options I see:
1. figure out the open questions there before Alpha
2. move the section to alternatives or notes (as a backup for the work already done) and add a graduation point about determining the remaining design points then

I would be leaning for (2.), any preference @alculquicondor ?

Reply:
2 seems fine. But leave notes for whatever work for jobset that is going to happen in alpha.

Reply:
Ok, moved the started API work to alternatives. I don't think we need any special work for JobSet in alpha then, except for this: https://github.com/kubernetes-sigs/kueue/pull/2725/files#diff-40027534eab1ffd244e3c960acbf91b55b9556bebd8cd56077ba5f3270648284R163-R167, which is already released, so we may consider dropping the note.